File: CH24/2014_pre_ec_p/calculator.go
package main

import (
    "flag"
    "fmt"
    "os"
    "bufio"
    "strings"
    "strconv"
    "math"
    "time"
)

func main() {
    t0 := time.Now()

    const (
        defaultFile = "film_logistics_input"
        usage       = "Input file"
    )

    var sourceFile string
    var movieCount int
    var firstlat float64
    var firstlon float64
    var prewlat float64
    var prewlon float64

    flag.StringVar(&sourceFile, "file", defaultFile, usage)
    flag.Parse()

    inputFile, err := os.Open("inputs/" + sourceFile + ".in");
    defer inputFile.Close()

    if err != nil {
        panic("Input file unreadable")
    }

    scanner  := bufio.NewScanner(bufio.NewReader(inputFile))
    i        := 0
    distance := 0.0

    for scanner.Scan() {

        if i == 0 {
            movieCount, _ = strconv.Atoi(scanner.Text())
        } else {
            coordinates := strings.Split(scanner.Text(), " ")
            lat, _      := strconv.ParseFloat(coordinates[0], 64)
            lon, _      := strconv.ParseFloat(coordinates[1], 64)

            if i == 1 {
                firstlat = lat
                firstlon = lon
            } else {
                distance = distance + getDistance(prewlat, prewlon, lat, lon)
            }

            if i == movieCount {
                distance = distance + getDistance(lat, lon, firstlat, firstlon)
            }

            prewlat = lat
            prewlon = lon
        }

        i = i + 1
    }

    writeToFile(sourceFile, distance)

    t1 := time.Now()
    fmt.Printf("The call took %v to run.\n", t1.Sub(t0))
    fmt.Println(distance / 1000)
}

func writeToFile(src string, dis float64) {
    source, _ := os.OpenFile("outputs/p" + src + ".out", os.O_APPEND | os.O_CREATE | os.O_WRONLY, 0777)
    source.Write([]byte(strconv.FormatFloat(dis, 'f', 6, 64)))
    source.Close()
}

func getDistance(lat1 float64, lon1 float64, lat2 float64, lon2 float64) (distance float64) {
    R      := 6371000.0
    sinLat := math.Sin(getRadian(lat1)) * math.Sin(getRadian(lat2))
    cosLat := math.Cos(getRadian(lat1)) * math.Cos(getRadian(lat2)) * math.Cos(getRadian(lon2-lon1))

    distance = math.Acos(sinLat + cosLat)  * R

    return distance
}

func getRadian(num float64) (rad float64) {
    rad = num * math.Pi / 180
    return  rad
}

File: CH24/2014_pre_ec_p/calculator.js
var startTime = +new Date(),
    argv      = process.argv;

if (3 > argv.length)
{
    console.log('Usage: node <filename.js> <input-file-id> [-v]');
    return;
}

var verbose = !!argv[3] && '-v' === argv[3],
    fs      = require('fs');

fs.readFile('inputs/' + argv[2] + '.in', processInput);

function processInput(err, data)
{
    if (err) { throw err; }

    var lines      = data.toString('utf-8').split("\n"),
        numCinemas = +lines.shift(),
        distance   = 0;

    for (
        var i = 1,
            coordinates1 = lines[0],
            idx1 = coordinates1.indexOf(' '),
            lat1 = toRad(+coordinates1.substring(0, idx1)),
            lon1 = toRad(+coordinates1.substring(idx1)),
            coordinates2, idx2, lat2, lon2;
        i <= numCinemas;
        i++
    ) {
        coordinates2 = lines[i];

        if (numCinemas === i)
        {
            coordinates2 = lines[0];
        }

        idx2 = coordinates2.indexOf(' ');
        lat2 = toRad(+coordinates2.substring(0, idx2));
        lon2 = toRad(+coordinates2.substring(idx2));

        distance += sphericalLawOfCosines(lat1, lon1, lat2, lon2);

        coordinates1 = coordinates2;
        idx1 = idx2;
        lat1 = lat2;
        lon1 = lon2;
    }

    processOutput(distance);
}

function processOutput(data)
{
    fs.writeFile('outputs/p' + argv[2] + '.out', data, function(err)
    {
        if (err) { throw err; }

        console.log(data + "\n");

        console.log('Elapsed time: ' + (+new Date() - startTime) + ' ms');
        console.log((process.memoryUsage().heapTotal / 1024 / 1024).toFixed(2) + ' MiB');
    });
}


function sphericalLawOfCosines(lat1, lon1, lat2, lon2)
{
    return Math.acos(Math.sin(lat1) * Math.sin(lat2) +
        Math.cos(lat1) * Math.cos(lat2) *
        Math.cos(lon2 - lon1)) * 6371;
}

function toRad(n)
{
    return n * Math.PI / 180;
}